prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -179,6 +192,121 @@ def filter_by_connected_components(lines, substations):
     return remaining_lines, remaining_substations
 
 
+def augment_line_voltages(
+    lines, substations, volt_class_defaults=None, state_threshold_100_161=0.95
+):
+    """Fill in voltages for lines with missing voltages, using a series of heuristics.
+    The ``lines`` dataframe will be modified in-place.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :param pandas.DataFrame substations: data frame of substations.
+    :param dict/pandas.Series volt_class_defaults: mapping of volt classes to default
+        replacement voltage values. If None, internally-defined defaults will be used.
+    :param int/float state_threshold_100_161: fraction of lines in the 100-161 kV range
+        which must be the same voltage for the most common voltage to be applied to
+        missing voltage lines in this state.
+    """
+
+    # Interpret input parameters
+    if volt_class_defaults is None:
+        volt_class_defaults = {
+            "UNDER 100": 69,
+            "220-287": 230,
+            "345": 345,
+            "500": 500,
+        }
+
+    # Replace dummy data with explicit 'missing'
+    lines.loc[lines.VOLTAGE == -999999, "VOLTAGE"] = pd.NA
+
+    # Set voltages based on voltage class defaults
+    null_voltages = lines.loc[lines.VOLTAGE.isna()]
+    replacement_voltages = null_voltages.VOLT_CLASS.map(volt_class_defaults)
+    lines.loc[lines.VOLTAGE.isna(), "VOLTAGE"] = replacement_voltages
+
+    # Use commonly-assigned voltages by state to fill 100-161 kV range
+    # Assume state grouping on one end is representative
+    missing_mask = lines.VOLTAGE.isna()
+    class_100_161_mask = lines.VOLT_CLASS == "100-161"
+    lines["STATE_1"] = lines.SUB_1_ID.map(substations.STATE)
+    state_100_161_lines = {
+        state: group.query("100 <= VOLTAGE <= 161").value_counts("VOLTAGE")
+        for state, group in lines.groupby("STATE_1")
+    }
```

Comment:
Would it be worth to create the same dictionary for `SUB_2_ID`, then combine the 2 dictionaries to boost the assignment of the 100-161 kV lines using this state-base method?

Reply:
So instead of matching the voltages just based on one arbitrary end's state, we would match voltages based on (state_from, state_to) tuples? I guess that could be helpful for states without a clear 'winning' voltage in this class, who may have a clear winning voltage for lines that cross a certain border. There may also be a risk that a state currently with a 'winning' voltage no longer passes the threshold when interior lines vs. intrastate lines are separated... but maybe that is a more reasonable assumption.

Reply:
I agree there is a chance that some states that are passing the threshold currently might not do so if we consider both ends. Not sure what method is better. Just throwing the idea.

Reply:
I like this change. The more information we are using, the closer we are approaching the truth, assuming these information is reliable. On the other hand, if it brings in ambiguity at the same time, then such ambiguity comes from the truth rather than our assumptions, which is also an improvement.

Reply:
Also, we could just substitute `SUB_1_ID` with `SUB_2_ID` to see how the assignment is changing. Just being curious.

Reply:
After running with a groupby on both (sorted, so that `(AL, TN)` and `(TN, AL)` are grouped together):
- After step 2, 12 more lines are mapped (7696 missing vs. 7708 before)
- After step 3, 1 fewer line is mapped (2119 missing vs. 2118 before)
- After step 4, the same number of lines are mapped (310 before and after)

Changes from multi-state grouping:
- Alabama to Florida is mapped to 115 kV, Alabama to Tennessee is mapped to 161 kV
- Arkansas to Louisiana is mapped to 115 kV, Arkansas to Missouri is mapped to 161 kV, Arkansas to Texas is mapped to 138 kV
- Arizona to California is mapped to 161 kV
- California to Nevada is mapped to 120 kV, California to Oregon is mapped to 115 kV
- Delaware to New Jersey and Delaware to Pennsylvania have no mapping (could have been auto-mapped to 138 kV before, based on Delaware grouping)
- Florida to Georgia is mapped to 115 kV
- Georgia to North Carolina is mapped to 161 kV, Georgia to South Carolina is mapped to 115 kV, Georgia to Tennessee is mapped to 161 kV
- Iowa to {Illinois, Minnesota, Missouri, Nebraska, Wisconsin} are mapped to 161 kV, Iowa to South Dakota is mapped to 115 kV
- Idaho to {Idaho, Montana, Oregon, Washington} are no longer mapped to anything (had been 138 kV), but Idaho to Nevada is mapped to 138 kV
- Illinois to Kentucky gets mapped to 161 kV, and Illinois to Missouri gets no mapping (Illinois was 138 kV previously, and other border states retain this mapping)
- Indiana to Kentucky gets no mapping (Indiana was 138 kV previously, and other border states retain this mapping)
- Kansas to Missouri gets 161 kV
- Kentucky to {Ohio, West Virginia} get 138 kV, Kentucky to Tennessee gets 161 kV
- Louisiana to Mississippi gets 115 kV, Louisiana to Texas gets 138 kV
- **Massachusetts to NY gets no mapping (had been 115 previously)**
- Maryland to {Virginia, West Virginia} get 138 kV
- Minnesota to North Dakota gets 115 kV (Minnesota previously had no mapping, but North Dakota was also 115 kV before)
- Missouri to Tennessee gets 161 kV
- Mississippi to Tennessee gets 161 kV
- Nebraska to South Dakota gets 115 kV (Nebraska previously had no mapping, but South Dakota was also 115 kV before)
- New Jersey to New York gets 115 kV, New Jersey to Pennsylvania gets 138 kV
- New Mexico to Utah gets nothing (New Mexico had gotten 115 kV before)
- **Oregon to Washington get no mapping, but both had been 115 kV before**
- Pennsylvania to West Virginia gets 138 kV
- Tennessee to Virginia gets 138 kV (all of Tennessee had gotten 161 kV before)
- Virginia to West Virginia gets 138 kV

Reply:
Overall, there are 326 lines whose voltages change between the one-state implementation and the two-state implementation.

Most of these differences are within states: Texas (46 lines), Indiana (32), Michigan (27), North Carolina (23), Alabama (21), Ohio (19), Arizona (19), New York (18), California (15), Illinois (13), Florida (12), Virginia (12), New Jersey (8), Louisiana (8), Nevada (7), Pennsylvania (6), Kentucky (5), Minnesota (5), South Carolina (5), Iowa (4), South Dakota (4), Kansas (3), Georgia (2), Tennessee (1), Oklahoma (1), New Mexico (1), North Dakota (1), and Wisconsin (1).

Only a handful are between states: New Jersey to Pennsylvania (2), Minnesota to South Dakota (1), Alabama to Florida (1), Maryland to Pennsylvania (1), Louisiana to Texas (1), Georgia to South Carolina (1).

Reply:
If we use the code from #210 to create buses, we see that using the two-state approach, there are 4,895 substations with multiple voltages, and 5,399 transformers (10,294 - 5,399) would be required to connect them:
```python
>>> from prereise.gather.griddata.hifld.data_process.transmission import build_transmission
>>> from prereise.gather.griddata.hifld.data_process.transmission import create_buses
>>> lines, substations = build_transmission()
>>> bus, bus2sub = create_buses(lines, substations)
>>> bus.groupby(bus2sub).count().baseKV.loc[bus.groupby(bus2sub).count().baseKV > 1]
sub_id
107655    2
107656    2
107658    2
107666    2
107674    2
         ..
310856    2
310861    2
310869    2
310882    2
310885    2
Name: baseKV, Length: 4895, dtype: int64
>>> bus.groupby(bus2sub).count().baseKV.loc[bus.groupby(bus2sub).count().baseKV > 1].sum()
10294
```

Using the existing one-state approach, we see 4,893 substations with multiple voltages, and 5,397 transformers would be required to connect them:
```python
>>> from prereise.gather.griddata.hifld.data_process.transmission import build_transmission
>>> from prereise.gather.griddata.hifld.data_process.transmission import create_buses
>>> lines, substations = build_transmission()
>>> bus, bus2sub = create_buses(lines, substations)
>>> bus.groupby(bus2sub).count().baseKV.loc[bus.groupby(bus2sub).count().baseKV > 1]
sub_id
107655    2
107656    2
107658    2
107666    2
107674    2
         ..
310856    2
310861    2
310869    2
310882    2
310885    2
Name: baseKV, Length: 4893, dtype: int64
>>> bus.groupby(bus2sub).count().baseKV.loc[bus.groupby(bus2sub).count().baseKV > 1].sum()
10290
```

Therefore, if we use the number of multi-voltage substations and the number of required connections within them as a quality metric, we see no significant regression, despite several hundred changes in individual line voltages.

Reply:
Changes made.

Reply:
Thanks for the detailed comparison